### PR TITLE
SMIBHID has a webserver! Closes #248

### LINKED ISSUE #248: Add a webserver to smibhid for admin purposes
For somakeit/smibhid#23 and somakeit/smib#241 we need a webserver to provide an admin interface. Issue is to investigate adding one to the async loop and see if it can be supported without first implementing MQTT.

### PR DESCRIPTION
**Pull request by sjefferson99**

### CODE REVIEW

File: smibhid/http/website.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
+from http.webserver import webserver
+from lib.ulogging import uLogger
+from lib.module_config import ModuleConfig
+from json import dumps
+import uasyncio
+
+class Web_App:
```
Comment (samcorky) [inline smibhid/http/website.py]: PEP8 naming violation - should be `WebApp`

File: smibhid/http/website.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    def create_api(self) -> None:
+        @self.app.route('/api')
+        async def api(request, response):
+            await response.send_file('/http/www/api.html')
+        
+        self.app.add_resource(wlan_mac, '/api/wlan/mac', wifi = self.wifi, logger = self.log)
+        self.app.add_resource(version, '/api/version', hid = self.hid, logger = self.log)
+    
+class wlan_mac():
```
Comment (samcorky) [inline smibhid/http/website.py]: again - should be PEP8

File: smibhid/http/website.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+    def get(self, data, wifi, logger: uLogger):
+        logger.info("API request - wlan/mac")
+        html = dumps(wifi.get_mac())
+        logger.info(f"Return value: {html}")
+        return html
+    
+class version():
```
Comment (samcorky) [inline smibhid/http/website.py]: PEP8

File: smibhid/http/webserver.py [hunk 4]
```diff
@@ -0,0 +1,692 @@
...
+
+
+class HTTPException(Exception):
+    """HTTP protocol exceptions"""
+
+    def __init__(self, code=400):
+        self.code = code
+
+
+class request:
```
Comment (samcorky) [inline smibhid/http/webserver.py]: PEP8

File: smibhid/http/webserver.py [hunk 5]
```diff
@@ -0,0 +1,692 @@
...
+            if ct == b'application/json':
+                return json.loads(data)
+            elif ct == b'application/x-www-form-urlencoded':
+                return parse_query_string(data.decode())
+        except ValueError:
+            # Re-generate exception for malformed form data
+            raise HTTPException(400)
+
+
+class response:
```
Comment (samcorky) [inline smibhid/http/webserver.py]: PEP8

File: smibhid/http/webserver.py [hunk 6]
```diff
@@ -0,0 +1,692 @@
...
+        else:
+            res_str = res
+        resp.add_header('Content-Type', 'application/json')
+        resp.add_header('Content-Length', str(len(res_str)))
+        resp.add_access_control_headers()
+        await resp._send_headers()
+        await resp.send(res_str)
+
+
+class webserver:
```
Comment (samcorky) [inline smibhid/http/webserver.py]: PEP8

**Review by samcorky — CHANGES_REQUESTED** [review]
Code is massive and looks OK, apart from some PEP8 naming convention violations.

I want to test this myself on my own SMIBHID dev board. Probably wont happen until tomorrow or Wednesday.

**Comment by sjefferson99** [comment]
PEP8 issues adjusted. The webserver module is not mine, but I have adjusted enough of it already to take on and continue adjusting things like class names and have adjusted these as well.

**Review by samcorky — COMMENTED** [review]
Happy with the PEP8 changes

**Comment by samcorky** [comment]
@sjefferson99 I want to test this tomorrow/Wednesday before I approve

**Comment by samcorky** [comment]
There is a merge conflict - please sort it :) @sjefferson99

**Review by samcorky — APPROVED** [review]
Tell my WiFi love her

**Pull request merged**